test/shared/src/main/scala/zio/test/experimental/TestAspect.scala

```diff
@@ -0,0 +1,378 @@
+   */
+  def dotty[LowerR, UpperR, LowerE, UpperE](
+    that: TestAspect[LowerR, UpperR, LowerE, UpperE]
+  ): TestAspect[LowerR, UpperR, LowerE, UpperE] =
+    if (TestVersion.isDotty) that else ???
+
+  /**
+   * An aspect that marks tests as ignored.
+   */
+  val dottyOnly: TestAspectAtLeastR[Annotations] =
+    if (TestVersion.isDotty) ??? else ignore
+
+  /**
+   * An aspect that retries a test until success, without limit.
+   */
+  val eventually: TestAspectAtLeastR[ZTestEnv] =
+    new TestAspect[Nothing, Any, Nothing, Any] {
+      type ModifyEnv[+R] = R with ZTestEnv
+      type ModifyErr[+E] = E
+
+      def some[R, E](predicate: String => Boolean, spec: ZSpec[R, E]): ZSpec[R with ZTestEnv, E] =
+        spec.transform[R with ZTestEnv, TestFailure[E], TestSuccess] {
+          case c @ Spec.SuiteCase(_, _, _) => c
+          case Spec.TestCase(label, test, annotations) =>
+            val debug: ZIO[R with ZTestEnv, TestFailure[E], TestSuccess] =
+              test.eventually
+            Spec.TestCase(label, if (predicate(label)) debug else test, annotations)
+        }
+    }
+
+  /**
+   * An aspect that retries a test until success, without limit.
+   */
+  def restore[R](service: R => Restorable): TestAspectAtLeastR[R] =
+    ??? //around(ZIO.accessM[R](r => service(r).save))(restore => restore)
+
+  /**
+   * An aspect that restores the
+   * [[zio.test.environment.TestClock TestClock]]'s state to its starting
+   * state after the test is run. Note that this is only useful when repeating
+   * tests.
+   */
+  def restoreTestClock: TestAspectAtLeastR[TestClock] =
+    restore[TestClock](_.get)
+
+  /**
+   * An aspect that restores the
+   * [[zio.test.environment.TestClock TestClock]]'s state to its starting
+   * state after the test is run. Note that this is only useful when repeating
+   * tests.
+   */
+  def restoreTestConsole: TestAspectAtLeastR[TestConsole] =
+    restore[TestConsole](_.get)
+
+  /**
+   * An aspect that restores the
+   * [[zio.test.environment.TestClock TestClock]]'s state to its starting
+   * state after the test is run. Note that this is only useful when repeating
+   * tests.
+   */
+  def restoreTestRandom: TestAspectAtLeastR[TestRandom] =
+    restore[TestRandom](_.get)
+
+  /**
+   * An aspect that restores the
+   * [[zio.test.environment.TestClock TestClock]]'s state to its starting
+   * state after the test is run. Note that this is only useful when repeating
+   * tests.
+   */
+  def restoreTestSystem: TestAspectAtLeastR[TestSystem] =
+    restore[TestSystem](_.get)
+
+  /**
+   * An aspect that restores the
+   * [[zio.test.environment.TestClock TestClock]]'s state to its starting
+   * state after the test is run. Note that this is only useful when repeating
+   * tests.
+   */
+  def restoreTestEnvironment: TestAspectAtLeastR[ZTestEnv] =
+    restoreTestClock >>>[Nothing, Any] restoreTestConsole >>>[Nothing, Any] restoreTestRandom >>>[Nothing, Any] restoreTestSystem
```

Comment:
We need to manually specify type parameters here.

Reply:
This one looks like the worst, although the `if` / `else` is pretty disappointing too.